File: benchmarks/auth-db-http-server/fastify/src/routes/database-route.mjs
import { databaseValue } from '../config.mjs';

export function registerDatabaseRoute({
  fastify,
  auth,
  baseUrl,
  database,
  authenticate,
}) {
  fastify.get('/bench/db', async (request, reply) => {
    const email = await authenticate({
      auth,
      baseUrl,
      nodeHeaders: request.headers,
    });
    if (email === null) {
      reply.code(401).type('application/json; charset=utf-8');
      return JSON.stringify({ error: 'unauthorized' });
    }

    const row = database
      .prepare('SELECT value FROM benchmark_values WHERE email = ?')
      .get(email);
    if (row?.value !== databaseValue) {
      reply.code(500).type('application/json; charset=utf-8');
      return JSON.stringify({ error: 'benchmark_row_missing' });
    }

    reply.type('application/json; charset=utf-8');
    return JSON.stringify({ email, value: row.value });
  });
}

File: benchmarks/auth-db-http-server/dart_edge/lib/src/routes/database_route.dart
import 'dart:convert';
import 'dart:io';

import 'package:dart_edge/dart_edge.dart';
import 'package:dart_edge_auth/dart_edge_auth.dart';
import 'package:dart_edge_sql/dart_edge_sql.dart';

import '../benchmark_config.dart';
import '../services.dart';

final class DatabaseRoute extends JsonRouteDefinition<Services, Object?> {
  @override
  RouteContract get contract => RouteContract(
    method: HttpMethod.get,
    path: '/bench/db',
    options: RouteOptions(
      operationId: 'benchmarkDatabase',
      success: ResponseSpec.json<Object?>(),
      errors: [ErrorResponse.unauthorized(code: 'unauthorized')],
    ),
  );

  @override
  Future<Object?> handle(RequestContext<Services> ctx) async {
    final email = ctx.requireAuthIdentity.email;
    if (email == null) {
      return ctx.res.code(HttpStatus.unauthorized).json({
        'error': 'unauthorized',
      });
    }

    final result = await ctx.services.database.execute(
      SqlStatement.named(
        'SELECT value FROM benchmark_values WHERE email = :email',
        {'email': email},
      ),
    );
    final value = result.single['value'];
    if (value is! String || value != benchmarkDatabaseValue) {
      return ctx.res.code(HttpStatus.internalServerError).json({
        'error': 'benchmark_row_missing',
      });
    }

    return ctx.res
        .type('application/json; charset=utf-8')
        .send(jsonEncode({'email': email, 'value': value}));
  }
}
